source/shared/cpp/ObjectModel/MarkDownParser.cpp

```diff
@@ -0,0 +1,477 @@
+    {
+        GetCh(ch);
+    }
+}
+
+bool MarkDownParser::IsLeftEmphasisDelimiter() 
+{
+    return ((m_emphasisState & ~InsideEmphasis) && 
+             m_delimiterCnts && 
+             !isspace(*m_curPos) && 
+             !(m_lookBehind == Alphanumeric && ispunct(*m_curPos)) && 
+             !(m_lookBehind == Alphanumeric && m_currentDelimiterType == Underscore));
+}
+
+// save left emphasis to look up table
+// return true, if left emphasis was detected and added
+bool MarkDownParser::PushLeftEmphasisToLookUpTableIfValid()
+{
+    // left emphasis detected, save emphasis for later reference
+    if (IsLeftEmphasisDelimiter())
+    {
+        m_leftLookUpTable.push_back(
+            Emphasis(m_delimiterCnts, m_currentWordIndex)); 
+        m_LeftEmphasisDetecting += m_delimiterCnts;
+        return true;
+    }
+    return false;
+}
+
+bool MarkDownParser::IsRightEmphasisDelimiter()
+{
+    if (isspace(*m_curPos) && 
+       (m_lookBehind != WhiteSpace) && 
+       (m_checkLookAhead || m_checkIntraWord || m_currentDelimiterType == Asterisk))
+    {        
+        return true;
+    }
+
+    if (isalnum(*m_curPos) && m_lookBehind != WhiteSpace)
+    {
+        if (!m_checkLookAhead && !m_checkIntraWord)
+        {
+            return true;
+        }
+
+        if (m_checkLookAhead || m_checkIntraWord)
+        {
+            return false;
+        }
+    }
+
+    if (ispunct(*m_curPos) && m_lookBehind != WhiteSpace)
+    {
+        if ((m_emphasisState & EmphasisEnd) || 
+           ((m_emphasisState & InsideEmphasis) && (m_curPos + 1 == m_text.end()))) 
+        {
+            return true;
+        }
+    }
+    return false;;
+}
+
+// save right emphasis to look up table
+// return true, if right emphasis was detected and added
+bool MarkDownParser::PushRightEmphasisToLookUpTableIfValid()
+{
+    if (IsRightEmphasisDelimiter())
+    {
+        if (m_LeftEmphasisDetecting)
+        {
+            m_rightLookUpTable.push_back(
+                Emphasis(m_delimiterCnts, m_currentWordIndex)); 
+            m_LeftEmphasisDetecting -= m_delimiterCnts;
+            return true;
+        }
+    }
+    return false;
+}
+
+void MarkDownParser::SetText(const std::string & txt)
+{
+    m_text = txt;
+}
+
+bool MarkDownParser::IsMarkDownDelimiter(char ch)
+{
+    return ((ch == '*' || ch == '_') && (m_lookBehind != Escape));
+}
+
+// Updates Emphasis MarkDown State
+void MarkDownParser::UpdateState()
+{
+    if (IsMarkDownDelimiter(*m_curPos))
+    {
+        m_emphasisState = (m_emphasisState & (OutsideEmphasis))?
+            EmphasisStart : EmphasisRun;
+    }
+    else
+    {
+        m_emphasisState = (m_emphasisState & InsideEmphasis)?
+            EmphasisEnd : EmphasisNone;
+    }
+
+}
+
+MarkDownParser::DelimiterType MarkDownParser::GetDelimiterTypeForCharAtCurrentPosition()
+{ 
+    return (*m_curPos == '*')? Asterisk : Underscore;
+}
+
+void MarkDownParser::StartNewTokenCapture()
+{
+    m_tokenizedString.push_back("");
```

Comment:
What does m_tokenizedString represent?  What does it mean to add an empty string to this vector?

Reply:
sorry, i will add comments. it contains tokens. adding empty string to it begins capturing new token. 

---
In reply to: [154743444](https://github.com/Microsoft/AdaptiveCards/pull/1003#discussion_r154743444) [](ancestors = 154743444)